### PR TITLE
Update linspace types

### PR DESCRIPTION
**Pull request by Enealor**
Changes the linspace functions to be more consistent as requested in #31991. The code has also been updated to avoid an early rounding error; the line `scalar_t step = (scalar_end - scalar_start) / static_cast<static_t>(steps-1)` can result in `step = 0` for integer scalars, and this gives unintended results. I examined the new output using
```
import torch

types = [torch.uint8, torch.int8, torch.short, torch.int, torch.long, torch.half, torch.float, torch.double]

print('Testing linspace:')
for type in types:
    print(type, torch.linspace(-2, 2, 10, dtype=type))
```
which returns
```
Testing linspace:
torch.uint8 tensor([254, 254, 254, 255, 255,   0,   0,   1,   1,   2], dtype=torch.uint8)
torch.int8 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2], dtype=torch.int8)
torch.int16 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2], dtype=torch.int16)
torch.int32 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2], dtype=torch.int32)
torch.int64 tensor([-2, -2, -2, -1, -1,  0,  0,  1,  1,  2])
torch.float16 tensor([-2.0000, -1.5557, -1.1113, -0.6670, -0.2227,  0.2227,  0.6660,  1.1113,
         1.5547,  2.0000], dtype=torch.float16)
torch.float32 tensor([-2.0000, -1.5556, -1.1111, -0.6667, -0.2222,  0.2222,  0.6667,  1.1111,
         1.5556,  2.0000])
torch.float64 tensor([-2.0000, -1.5556, -1.1111, -0.6667, -0.2222,  0.2222,  0.6667,  1.1111,
         1.5556,  2.0000], dtype=torch.float64)
```
which is the expected output: `uint8` overflows as it should, and the result of casting from a floating point to an integer is correct.

This PR does not change the logspace function.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 6f73f2e9:


**None of the build failures appear to be your fault**.

* **1/3** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=4dce482acb2f0b248e4886b3069dca8e3a1b7681&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 4dce482a since Jan 15
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **2/3** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6f73f2e91bb9031e045e5831b305d0a3b53d0a81).

### :snowflake: 2 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4253102) pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86358976) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86358976)) :snowflake:
<details>
<summary>
<code>Jan 15 09:26:00   Error generating file</code>
</summary>

```
Jan 15 09:26:00 /var/lib/jenkins/workspace/c10/util/ArrayRef.h(278): warning: attribute does not apply to any entity 
Jan 15 09:26:00  
Jan 15 09:26:00 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/RangeFactories.cu(35): error: no operator "*" matches these operands 
Jan 15 09:26:00             operand types are: const scalar_t * int 
Jan 15 09:26:00  
Jan 15 09:26:00 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/RangeFactories.cu(35): error: no operator "*" matches these operands 
Jan 15 09:26:00             operand types are: const scalar_t * int 
Jan 15 09:26:00  
Jan 15 09:26:00 2 errors detected in the compilation of "/tmp/tmpxft_000001de_00000000-6_RangeFactories.cpp1.ii". 
Jan 15 09:26:00 CMake Error at torch_cuda_generated_RangeFactories.cu.o.Release.cmake:281 (message): 
Jan 15 09:26:00   Error generating file 
Jan 15 09:26:00   /var/lib/jenkins/workspace/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_RangeFactories.cu.o 
Jan 15 09:26:00  
Jan 15 09:26:00  
Jan 15 09:26:00 caffe2/CMakeFiles/torch_cuda.dir/build.make:840: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_RangeFactories.cu.o' failed 
Jan 15 09:26:00 make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_RangeFactories.cu.o] Error 1 
Jan 15 09:26:00 make[2]: *** Waiting for unfinished jobs.... 
Jan 15 09:26:01 /var/lib/jenkins/workspace/c10/util/ArrayRef.h(278): warning: attribute does not apply to any entity 
Jan 15 09:26:01  
Jan 15 09:26:01 /var/lib/jenkins/workspace/c10/util/ArrayRef.h(278): warning: attribute does not apply to any entity 
Jan 15 09:26:01  
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4253096) pytorch_libtorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86361338) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86361338)) :snowflake:
<details>
<summary>
<code>Jan 15 09:28:43   Error generating file</code>
</summary>

```
Jan 15 09:28:42 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/RangeFactories.cu(35): error: no operator "*" matches these operands 
Jan 15 09:28:42             operand types are: const scalar_t * int 
Jan 15 09:28:42  
Jan 15 09:28:42 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/RangeFactories.cu(35): error: no operator "*" matches these operands 
Jan 15 09:28:42             operand types are: const scalar_t * int 
Jan 15 09:28:42  
Jan 15 09:28:42 2 errors detected in the compilation of "/tmp/tmpxft_00007d56_00000000-6_RangeFactories.cpp1.ii". 
Jan 15 09:28:42 -- Removing /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_RangeFactories.cu.o 
Jan 15 09:28:43 /usr/bin/cmake -E remove /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_RangeFactories.cu.o 
Jan 15 09:28:43 CMake Error at torch_cuda_generated_RangeFactories.cu.o.Debug.cmake:281 (message): 
Jan 15 09:28:43   Error generating file 
Jan 15 09:28:43   /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_RangeFactories.cu.o 
Jan 15 09:28:43  
Jan 15 09:28:43  
Jan 15 09:28:43 caffe2/CMakeFiles/torch_cuda.dir/build.make:882: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_RangeFactories.cu.o' failed 
Jan 15 09:28:43 make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_RangeFactories.cu.o] Error 1 
Jan 15 09:28:43 make[2]: *** Waiting for unfinished jobs.... 
Jan 15 09:28:43 -- Generating temporary cmake readable file: /var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_FractionalMaxPool2d.cu.o.depend.tmp 
Jan 15 09:28:43 /usr/bin/cmake -D input_file:FILEPATH=/var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_FractionalMaxPool2d.cu.o.NVCC-depend -D output_file:FILEPATH=/var/lib/jenkins/cpp-build/caffe2/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_FractionalMaxPool2d.cu.o.depend.tmp -D verbose=1 -P /var/lib/jenkins/workspace/cmake/Modules_CUDA_fix/upstream/FindCUDA/make2cmake.cmake 
Jan 15 09:28:43 CMake Warning at /var/lib/jenkins/workspace/cmake/Modules_CUDA_fix/upstream/FindCUDA/make2cmake.cmake:76 (message): 
Jan 15 09:28:43    Removing non-existent dependency file: TH/generic/THStorage.h 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by Enealor** [comment]
Status: flabbergasted that the CUDA builds are saying multiplication via `*` isn't supported between `scalar_t` and `int` types, as that was used in the previous build. My concern now is that changing just the CPU will result in discrepancies in output (namely, the CUDA version outputting garbage in the case that `dtype=torch.long`). However, I don't currently have the ability to build and test the CUDA portion locally.

CircleCI says that it may be due to flaky builds, but it's pretty wild that it supports multiplication used this way sometimes and not others.

**Comment by ngimel** [comment]
Awesome! Can you please find where the linspace is tested (likely test_torch.py) and add tests for the different types there?  A wiki page on writing the tests is here https://github.com/pytorch/pytorch/wiki/Writing-tests-that-run-on-all-available-device-types, likely you don't need a new test, you just need to modify an existing one. There are @dtypesIfCPU and @dtypesIfCUDA decorators available to test different types on the different devices, but when using those you also have to specify @dtypes to make sure that XLA tests are also covered, like this:
```
 @dtypes(torch.float, torch.bfloat16, torch.long, torch.bool)
 @dtypesIfCPU(torch.float, torch.long, torch.bool, torch.bfloat16)
 @dtypesIfCUDA(torch.half, torch.long, torch.bool)
```
(don't pay attention to actual types, I copy-pasted, put in what you think is needed).

**Comment by Enealor** [comment]
Okay. Thank you for the direction. I will add those tests soon. However, I noticed issues while creating the test, and I'm going to add those commits first. Namely,
* I found that `torch.half` equality checking is not available for CPU tensors, and so I plan on removing this type from the dispatch.
* I realized that the results above are always rounded down rather than truncated. As I expect the latter (and NumPy produces the same), I plan on fixing this. There will be a discrepancy for `torch.uint8` however, as 'truncated' and 'rounded down' are the same in this case due to lack of negatives. The way I plan on fixing this is by making it so that a separate dispatch is called for floating and integral types.
* Because of the last change and the issues leading up to it, I think I figured out how to add complex types to `linspace_cuda` with a minor code change. However, I don't know how to test this. It looks like they're not fully supported in PyTorch yet (based on [this issue](https://github.com/pytorch/pytorch/issues/755), at least.)

**Comment by Enealor** [comment]
I have added tests, which I will summarize below. The CUDA version of `linspace` is failing the new tests as it is not producing the expected results. I believe rounding error is occurring somewhere, I am just not yet sure where. Only the `torch.half` version of the test is still passing.

I still haven't determined why the XLA build is failing. The error being given corresponds to changes from torch_xla that I do not understand yet.

The tests:
* I added `dtype` to the function parameters, with the decorator `@dtypes(torch.int8, torch.short, torch.int, torch.long, torch.float, torch.double)`. I also used the CUDA decorator to add these types along with `torch.half`.
* I modified the other `linspace` function calls to add `dtype`.
* I added a check against a simple shorthand of doing the same in python, which functions as a comparison point. The result of `linspace(-100, 100, 301, device=device, dtype=dtype)` is compared to the result of `torch.tensor([-100. + (2. / 3.) * i for i in range(301)], device=device, dtype=torch.double).to(dtype)`.

Currently, this last test is spot on for the CPU. On the GPU, it fails as I mentioned above. I am working on determining why.

File: aten/src/ATen/native/cuda/RangeFactories.cu [hunk 1]
```diff
@@ -31,18 +31,29 @@ Tensor& linspace_cuda_out(Tensor& result, Scalar start, Scalar end, int64_t step
...
+        scalar_t val = scalar_start + step * static_cast<scalar_t>(ind);
+        return val;
+      });
+    });
   } else {
     AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, r.scalar_type(), "linspace_cuda", [&]() {
       scalar_t scalar_start = start.to<scalar_t>();
       scalar_t scalar_end = end.to<scalar_t>();
-      scalar_t step = (scalar_end - scalar_start) / static_cast<scalar_t>(steps - 1);
+      double step = static_cast<double>(scalar_end - scalar_start) / (steps - 1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/RangeFactories.cu]: @enearlor can you please leave a comment here about why double is needed (is it?). Operations with doubles are typically much slower on cuda, so unless absolutely necessary we try not to use them, and it seems like in most cases linspace should be ok with a float step. I know you've ran many experiments, so I'm curious to hear which cases fail with floats.

Comment (Enealor) [reply aten/src/ATen/native/cuda/RangeFactories.cu]: I chose to use `double` after finding that I was having issues with the numerical accuracy of `step` when `linspace` is dispatched to `float`. The inaccuracy is small, and is typically around the type's `eps`. I reasoned that the discrepancy is due to Python doing everything in higher precision by default, and then rounding at the end. Replicating this in `linspace` fixed the discrepancy.

I considered the performance hit this might cause, but I have not been able to investigate it as I can't directly test the CUDA build. I reasoned that the change in overall clock time would be minor as `linspace` is not frequently recalled. If this is a concern, then I can change it back. I list my suggestions below.

At the moment, these are the solutions I'm considering.
* Leave the code as-is, other then the corrections for test_linspace and the XLA build. This ideally results in the CUDA and CPU versions giving the same results.
* Change the type back to `scalar_t`. This would leave minor numerical discrepancies between the output of the CUDA and CPU version, but it would keep the CUDA version performing at the speed it is currently. This would require me to separate out the Integral form though, as an integral `step` can be truncated to 0. This would also mean that I'd have to allow for error between the expected result and actual result in the tests, but I believe that is the current state of affairs as well.

Comment (ngimel) [reply aten/src/ATen/native/cuda/RangeFactories.cu]: I'd prefer changing the type back to `scalar_t` and increasing test tolerance if needed. Separating integral dispatch is fine. Thank you!

Comment (Enealor) [reply aten/src/ATen/native/cuda/RangeFactories.cu]: Okay. I added a commit to address this. The integral dispatch still uses `float` for the `step`, but otherwise uses `scalar_t`.

To verify, is the CPU version as is? It relies on declaring the step to be `double` type, as this did prior to the commit.

Comment (ngimel) [reply aten/src/ATen/native/cuda/RangeFactories.cu]: Yes, for CPU doubles don't carry such a performance penalty. XLA failures are probably unrelated, can you try rebasing and see it they are fixed?

**Comment by Enealor** [comment]
The CUDA tests are passing. However, I deliberately chose an easy case to increase the likelihood of passing, and I'd like feedback on that. 

I changed the test to use `linspace(-100, 100, 401, device=device, dtype=dtype)`. This results in `step = 0.5`, which is easily represented. The previous test effectively used `step = 2. / 3.`, which is impacted by the rounding. Because of this change, the minor errors that cropped up for integers due to rounding do not appear. Additionally, I allowed for a tolerance depending on `dtype` and `device`.

Outside of this case, the result of calling `linspace` with an integral `dtype` will not be consistent across devices due to rounding. However, the differences were infrequent for the values I tested.

File: aten/src/ATen/native/cuda/RangeFactories.cu (hunk 1 above)
Comment (Enealor) [reply aten/src/ATen/native/cuda/RangeFactories.cu]: The XLA failures appear to be fixed after rebasing. The onnx build is failing for potentially similar reasons.

**Comment by ngimel** [comment]
@enealor is this ready for landing?

**Comment by Enealor** [comment]
Outside of the testing strategy questions I had in the last post and the onnx build (which may be unrelated), yes.

**Comment by ngimel** [comment]
@enealor, your testing strategy is fine, and we can live with minor differences on cpu and gpu for integer indexing (after all, there was a bug in integer indexing on the cpu that you fixed, and it was not supported on the gpu at all). Thank you!

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
@enealor can you please rebase? pytorchbot refuses to do that for me.
